_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,644 @@
+
+- RedirectURI: **(pre-filled)**
+- Alias: **auth0**
+- User discovery endpoint: **On**
+- Discovery endpoint: **https://\<auth0-domain\>/.well-known/openid-configuration**
+- Client authentication: **Client secret sent as post**
+- Client ID: **auth0-client-id**
+- Client Secret: **auth0-client-secret**
+
+Click on **Add** to continue the configuration. Below the Client Secret field, click on **Advanced**. In the _Scopes_ field, set `openid profile email offline_access` and click on **Save**.
+
+On the left menu, choose **Authentication**. In the flows table, choose **browser**. In the _Identity Provider Redirector_ step, click the **gear** icon. Set an alias for the configuration, and set **auth0** as the default identity provider. This configuration will skip the Keycloak sign-in form, and display the Auth0 sign-in form directly.
+
+{% img blog/keycloak-kubernetes-prod/auth0-redirector.png alt:"Identity provider redirector to Auth0" width:"450" %}{: .center-image }
+
+Go back to **Authentication**, and choose the **first broker login** flow. Disable the _Review Profile_ step.
+
+### Map the roles claim
+
+The users that sign in through Auth0 are imported to Keycloak. The role assigned in Auth0 must be mapped to the pre-configured roles in Keycloak for the authorization mechanism to work. In the Keycloak Administration Console, go to **Identity providers** again, and choose **auth0**. Then choose the **Mappers** tab. Click **Add mapper**. Fill in the values as follows and click **Save**:
+
+- Name: **ROLE_USER**
+- Sync mode override: **Force** (update the user roles during every login)
+- Mapper type: **Claim to Role**
+- Claim: **https://www\\.jhipster\\.tech/roles** (escape the dot `.` with backslash `\.`)
+- Claim Value: **ROLE_USER**
+- Role: Select **ROLE_USER**
+
+Repeat the process for mapping ROLE_ADMIN.
+
+{% img blog/keycloak-kubernetes-prod/role-mapper.png alt:"Add mapper form" width:"800" %}{: .center-image }
+
+### Test Keycloak as an identity broker
+
+Navigate to **http://gateway.\<namespace\>.\<public-ip\>.nip.io**. The gateway homepage should display. Click on **sign in** and you should be redirected to the Auth0 sign-in form:
+
+{% img blog/keycloak-kubernetes-prod/auth0-login.png alt:"Add mapper form" width:"350" %}{: .center-image }
+
+After signing in, the gateway home page will display the username (email). If roles were mapped correctly, the user should be able to create entities (granted by ROLE_USER), and the Administration menu will also display (granted by ROLE_ADMIN).
+
+## Learn More about Keycloak in production
+
+I hope you enjoyed this post and learned about some best practices for deploying Keycloak to production when doing JHipster development. Keep learning about Keycloak and Spring Boot! Check out the following links:
```

Comment:
I think we should have two additional sections to make this a "best practices" post.

1. Show how to change the admin's password after everything is first deployed. 
2. The apps aren't using HTTPS. Will the HTTPS certs eventually be valid? If so, I think we should mention it. If not, we should add info on how to make them HTTPS or link to a guide that shows how.

Forcing HTTPS might be a good idea too. JHipster's documentation shows how to do this. https://www.jhipster.tech/security/#https

Reply:
We should also show how to destroy the cluster so people don't forget. 

```
gcloud container clusters delete jhipster-cluster --zone=southamerica-east1-a
gcloud compute addresses delete gateway-ingress-ip --global
```

Reply:
> We should also show how to destroy the cluster so people don't forget.
> 
> ```
> gcloud container clusters delete jhipster-cluster --zone=southamerica-east1-a
> gcloud compute addresses delete gateway-ingress-ip --global
> ```

Added the commands

Reply:
> I think we should have two additional sections to make this a "best practices" post.
> 
> 1. Show how to change the admin's password after everything is first deployed.
> 2. The apps aren't using HTTPS. Will the HTTPS certs eventually be valid? If so, I think we should mention it. If not, we should add info on how to make them HTTPS or link to a guide that shows how.
> 
> Forcing HTTPS might be a good idea too. JHipster's documentation shows how to do this. https://www.jhipster.tech/security/#https

The was narrowing the focius on best practices for deploying Keycloak but I see the point. 

I will add point 1 to the post.

As the certificate for Kecloak TLS is issued by the Let’s Encrypt staging environment, the browser won’t trust it by default. This is mentioned in the post. I did not test with Let's Encrypt production. I am assuming certificates will be trusted by browsers if issued by Let's Encrypt production for a real domain.

I will add info on HTTPS for the apps.

Reply:
Added info on HTTPS for the gateway.